make test_cover failure

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions.-->

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as you can.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner. We might ask you to provide additional logs and data (tendermint & app)
in a case of bug.
-->

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): v0.17.0


**ABCI app** (name for built-in, URL for self-written if it's publicly available):


**Merkleeyes version** (use `git rev-parse --verify HEAD`, skip if you don't use it):


**Environment**:
- **OS** (e.g. from /etc/os-release): macos high sierra
- **Install tools**: source
- **Others**:


**What happened**:
Failed followed cases:
- TestReactorVotingPowerChange
- TestReactorValidatorSetChanges
- TestWALCrash

**What you expected to happen**:


**How to reproduce it** (as minimally and precisely as possible):
make test_cover
**Logs (you can paste a part showing an error or attach the whole file)**:

> === RUN   TestReactorVotingPowerChange
> D[04-04|09:58:07.423] ---------------------------- Testing changing the voting power of one validator a few times 
> --- FAIL: TestReactorVotingPowerChange (10.25s)
> === RUN   TestReactorValidatorSetChanges
> I[04-04|09:58:19.469] ---------------------------- Testing adding one validator 
> I[04-04|09:58:34.721] ---------------------------- Testing changing the voting power of one validator 
> I[04-04|09:58:51.288] ---------------------------- Testing adding two validators at once 
> I[04-04|09:59:08.884] ---------------------------- Testing removing two validators at once 
> --- FAIL: TestReactorValidatorSetChanges (73.00s)
> === RUN   TestReactorWithTimeoutCommit
> --- PASS: TestReactorWithTimeoutCommit (0.36s)
> === RUN   TestWALCrash
> === RUN   TestWALCrash/empty_block
> === RUN   TestWALCrash/block_with_a_smaller_part_size
> === RUN   TestWALCrash/many_non-empty_blocks

**`/dump_consensus_state` output for consensus bugs**

**Anything else do we need to know**:

@melekes 
--- FAIL: TestReactorVotingPowerChange (10.25s)
        testing.go:699: race detected during execution of test
I[04-04|09:58:16.808] Stopping baseWAL                             module=consensus validator=1 wal=/home/ubuntu/.tendermint_test/consensus_voting_power_changes_test_1/data/cs.wal/wal impl=baseWAL
E[04-04|09:58:16.805] Stopping peer for error                      module=p2p validator=3 peer="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" err=EOF
D[04-04|09:58:16.808] Stopping Peer (ignoring: already stopped)    module=p2p validator=3 peer=pipe impl="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}"
D[04-04|09:58:16.805] setHasVote                                   module=consensus validator=3 peerH/R=14/0 H/R=13/0 type=2 index=3
I[04-04|09:58:16.809] enterPrevote: ProposalBlock is valid         module=consensus validator=2 height=14 round=0
D[04-04|09:58:16.809] Sending vote message                         module=consensus validator=3 ps="PeerState{\n  Key   b07f24cf7740116aff2d751b2b3668f3503ff9ff\n  PRS   PeerRoundState{\n    14/0/RoundStepNewHeight @2018-04-04 09:58:16.781386814 +0000 UTC m=+18.378220154\n    Proposal 0:000000000000 -> nil-BitArray\n    POL      nil-BitArray (round -1)\n    Prevotes   nil-BitArray\n    Precommits nil-BitArray\n    LastCommit BA{4:X__X} (round 0)\n    Catchup    nil-BitArray (round -1)\n  }\n  Stats peerStateStats{votes: 13, blockParts: 8}\n}" vote="Vote{3:FF51E1CF1D4F 13/00/2(Precommit) BECE8F5E42FF {/C2B15B3877B1.../} @ 2018-04-04T09:58:16.414Z}"
D[04-04|09:58:16.809] No votes to send, sleeping                   module=consensus validator=3 peer="Peer{MConn{pipe} b07f24cf7740116aff2d751b2b3668f3503ff9ff in}" rs.Height=14 prs.Height=14 localPV=BA{4:____} peerPV=nil-BitArray localPC=BA{4:____} peerPC=nil-BitArray
D[04-04|09:58:16.806] Received bytes                               module=p2p validator=2 peer=pipe chID=32 msgBytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0002\u0001\u0003"
D[04-04|09:58:16.809] Receive                                      module=consensus validator=2 src="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" chId=32 bytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0002\u0001\u0003"
D[04-04|09:58:16.809] Read Msg Packet                              module=p2p validator=2 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:01000000000000000E00030000 T:1}"
D[04-04|09:58:16.809] Received bytes                               module=p2p validator=2 peer=pipe chID=32 msgBytes="\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u000e\u0000\u0003\u0000\u0000"
D[04-04|09:58:16.810] Receive                                      module=consensus validator=2 src="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" chId=32 bytes="\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u000e\u0000\u0003\u0000\u0000"
D[04-04|09:58:16.810] Read Msg Packet                              module=p2p validator=2 peer=pipe conn=MConn{pipe} packet="MsgPacket{22:14010114FF51E1CF1D4F263046A2EF6C57EAF5DCCB3243EF0103000000000000000D001522332DF0F85380020114BECE8F5E42FF001E95E143135C3FC7EAC24639D901010114688E19C5D3A709460F916CC7ADE84E1D1BEE4CDD01C2B15B3877B143982CC35AD1A7CA95851242DC1160CA7BF5D2F108C11AFA34F8C77CA098784995C1AF0E95E01A8B7028904A77D1B0D1B465DA81B42D965F700B T:1}"
D[04-04|09:58:16.810] Received bytes                               module=p2p validator=2 peer=pipe chID=34 msgBytes="\u0014\u0001\u0001\u0014\ufffdQ\ufffd\ufffd\u001dO&0F\ufffd\ufffdlW\ufffd\ufffd\ufffd\ufffd2C\ufffd\u0001\u0003\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0015\"3-\ufffd\ufffdS\ufffd\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\ufffdN\u00
01d\u001b\ufffdL\ufffd\u0001±[8w\ufffdC\ufffd,\ufffdZѧ?~U\ufffd\u0012B\ufffd\u0011`\ufffd{\ufffd\ufffd\ufffd\u0008\ufffd\u001a\ufffd4\ufffd\ufffd|\ufffd\ufffdxI\ufffd\ufffd\ufffd\u000e\ufffd\ufffd\u001a\ufffdp(\ufffdJwѰѴe?~A\ufffd-\uff
ffd_p\u000b"
D[04-04|09:58:16.810] Receive                                      module=consensus validator=2 src="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" chId=34 bytes="\u0014\u0001\u0001\u0014\ufffdQ\ufffd\ufffd\u001dO&0F\ufffd\ufffdlW\ufffd\ufffd\ufffd\ufffd2C\ufffd\u0001\u0003\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0015\"3-\ufffd\ufffdS\ufffd\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\ufffdN\u001d\u001b\ufffdL\ufffd\u0001±[8w\ufffdC\ufffd,\ufffdZѧ?~U\ufffd\u0012B\ufffd\u0011`\ufffd{\ufffd\ufffd\ufffd\u0008\ufffd\u001a\ufffd4\ufffd\ufffd|\ufffd\ufffdxI\ufffd\ufffd\uff
ffd\u000e\ufffd\ufffd\u001a\ufffdp(\ufffdJwѰѴe?~A\ufffd-\ufffd_p\u000b" 
D[04-04|09:58:16.810] Read Msg Packet                              module=p2p validator=2 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:16000000000000000D00020114BECE8F5E42FF001E95E143135C3FC7EAC24639D901010114688E19C5D3A709460F916CC7ADE84E1D1BEE4CDD T:1}"
D[04-04|09:58:16.810] Received bytes                               module=p2p validator=2 peer=pipe chID=32 msgBytes="\u0016\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\ufffdN\u001d\u001b\ufffdL\ufffd" 
D[04-04|09:58:16.810] Receive                                      module=consensus validator=2 src="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" chId=32 bytes="\u0016\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\ufffdN\u001d\u001b\ufffdL\ufffd" 
D[04-04|09:58:16.810] Read Msg Packet                              module=p2p validator=2 peer=pipe conn=MConn{pipe} packet="MsgPacket{22:14010114A047ADAEB2EA71D2C91A3188339F1B363EDDAD850101000000000000000D001522332DEF995FC0020114BECE8F5E42FF001E95E143135C3FC7EAC24639D901010114688E19C5D3A709460F916CC7ADE84E1D1BEE4CDD0147A183CDBC56C0CA1341AB94D721FF64CF4F468C017B2FAC4C9EB0F0C2DA5FE575963C3490D60B33BFA10C8028FFF0CA8E1DC70229872CB3E80AF4BB88273802 T:1}"
D[04-04|09:58:16.811] Received bytes                               module=p2p validator=2 peer=pipe chID=34 msgBytes="\u0014\u0001\u0001\u0014\ufffdG\ufffd\ufffd\ufffd\ufffdq\ufffd\ufffd\u001a1\ufffd3\ufffd\u001b6>ݭ\ufffd\u0001\u0001\u000
00\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0015\"3-\ufffd\ufffd_\ufffd\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\uffff
dN\u001d\u001b\ufffdL\ufffd\u0001G\ufffd\ufffdͼV\ufffd\ufffd\u0013A\ufffd\ufffd\ufffd!\ufffdd\ufffdOF\ufffd\u0001{/\ufffdL\ufffd\ufffd\ufffd\ufffd\ufffd_\ufffdu\ufffd<4\ufffd\ufffd\u000b3\ufffd\ufffd\u000c\ufffd(\ufffd\ufffd?~N\u001d\ufff
fd\u0002)\ufffd,\ufffd\ufffd\n\ufffd\ufffd\ufffd'8\u0002"
D[04-04|09:58:16.811] Receive                                      module=consensus validator=2 src="Peer{MConn{pipe} 07fbfd19d2963871b5c3a8b206bf6663a751303b in}" chId=34 bytes="\u0014\u0001\u0001\u0014\ufffdG\ufffd\ufffd\ufffd\ufffdq\ufffd\ufffd\u001a1\ufffd3\ufffd\u001b6>ݭ\ufffd\u0001\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\r\u0000\u0015\"3-\ufffd\ufffd_\ufffd\u0002\u0001\u0014\ufffd?~O^B\ufffd\u0000\u001e\ufffd\ufffdC\u0013\\?\ufffd\ufffd\ufffdF9\ufffd\u0001\\
u0001\u0001\u0014h\ufffd\u0019\ufffdӧ\tF\u000f\ufffdlǭ\ufffdN\u001d\u001b\ufffdL\ufffd\u0001G\ufffd\ufffdͼV\ufffd\ufffd\u0013A\ufffd\ufffd\ufffd!\ufffdd\ufffdOF\ufffd\u0001{/\ufffdL\ufffd\ufffd\ufffd\ufffd\ufffd_\ufffdu\ufffd<4\ufffd\uu
fffd\u000b3\ufffd\ufffd\u000c\ufffd(\ufffd\ufffd?~N\u001d\ufffd\u0002)\ufffd,\ufffd\ufffd\n\ufffd\ufffd\ufffd'8\u0002"
E[04-04|09:58:16.811] MConnection flush failed                     module=p2p validator=3 peer=pipe err="io: read/write on closed pipe"
D[04-04|09:58:16.811] Send Pong                                    module=p2p validator=2 peer=pipe
D[04-04|09:58:16.811] Flush                                        module=p2p validator=2 peer=pipe conn=MConn{pipe}
E[04-04|09:58:16.811] MConnection flush failed                     module=p2p validator=2 peer=pipe err="io: read/write on closed pipe"
I[04-04|09:58:16.812] Stopping baseWAL                             module=consensus validator=3 wal=/home/ubuntu/.tendermint_test/consensus_voting_power_changes_test_3/data/cs.wal/wal impl=baseWAL

--- FAIL: TestReactorValidatorSetChanges (73.00s)
        testing.go:699: race detected during execution of test
I[04-04|09:59:29.817] Stopping queryMaj23Routine for peer          module=consensus validator=6 peer="Peer{MConn{pipe} fb529f8961d429da653af361e49b46f8cc9e6e79 in}"
D[04-04|09:59:29.817] setHasVote                                   module=consensus validator=5 peerH/R=18/0 H/R=17/0 type=2 index=0
D[04-04|09:59:29.818] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{22:14010114A6DF319BFB4BC844C651AF20F59395DA5A8676A301010000000000000011001522333DCC0BDC80020114629B77BE310C25AE34A893A93D61AFF204018A2801010114C81BDD2EF2B3541B6DFB667EDA3843297B42C3E30179FFF77E25721D8999C18EB71773C4635FB83C5BBA41934B5299B5FAC33F81E89D4C012E2E02EE2F125E7D11EE8F239F2D2AC22B881C329725F073C3C07CFD01 T:1}"
D[04-04|09:59:29.818] Received bytes                               module=p2p validator=5 peer=pipe chID=34 msgBytes="\u0014\u0001\u0001\u0014\ufffd\ufffd1\ufffd\ufffdK\ufffdD\ufffdQ\ufffd \ufffd\ufffd\ufffd\ufffdZ\ufffdv\ufffd\u0001\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0015\"3=\ufffd\u000b?~@\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd\u0001y\ufffd\ufffd~%r\u001d\ufffd\ufffd\ufffd\ufffd\ufffd\u0017s\ufffdc_\ufffd<[\ufffdA\ufffdKR\ufffd\ufffd\ufffd\ufffd?\ufffd\ufffd\ufffdL\u0001..\u0002\ufffd/\u0012^}\u0011\ufffd\ufffd#\ufffd-*\ufffd+\ufffd\u001c2\ufffd%\ufffds\ufffd\ufffd|\ufffd\u0001"
D[04-04|09:59:29.818] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} 7a8b341c9adabdacfae9053c4cff036ea5322bd6 in}" chId=34 bytes="\u0014\u0001\u0001\u0014\ufffd\ufffd1\ufffd\ufffdK\ufffdD\ufffdQ\ufffd \ufffd\ufffd\ufffd\ufffdZ\ufffdv\ufffd\u0001\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0015\"3=\ufffd\u000b?~@\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd\u0001y\ufffd\ufffd~%r\u001d\ufffd\ufffd\ufffd\ufffd\ufffd\u0017s\ufffdc_\ufffd<[\ufffdA\ufffdKR\ufffd\ufffd\ufffd\ufffd?\ufffd\ufffd\ufffdL\u0001..\u0002\ufffd/\u0012^}\u0011\ufffd\ufffd#\ufffd-*\ufffd+\ufffd\u001c2\ufffd%\ufffds\ufffd\ufffd|\ufffd\u0001"
D[04-04|09:59:29.818] setHasVote                                   module=consensus validator=5 peerH/R=18/0 H/R=17/0 type=2 index=0
I[04-04|09:59:29.814] Stopping Peer                                module=p2p validator=6 peer=pipe impl="Peer{MConn{pipe} fb529f8961d429da653af361e49b46f8cc9e6e79 in}"
D[04-04|09:59:29.819] Stopping MConnection (ignoring: already stopped) module=p2p validator=6 peer=pipe impl=MConn{pipe}
D[04-04|09:59:29.814] Received bytes                               module=p2p validator=5 peer=pipe chID=32 msgBytes="\u0002\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0001\u0001\u0001\u0014~\u001a\ufffd\ufffd\ufffd_j\ufffd\ufffdf\ufffd\u001e\u000f?~C\ufffd\ufffdl\ufffd>\u0001\u0001\u0001\u0001\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0001"
D[04-04|09:59:29.819] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=32 bytes="\u0002\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0001\u0001\u0001\u0014~\u001a\ufffd\ufffd\ufffd_j\ufffd\ufffdf\ufffd\u001e\u000f?~C\ufffd\ufffdl\ufffd>\u0001\u0001\u0001\u0001\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0001"
D[04-04|09:59:29.819] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:16000000000000001100020114629B77BE310C25AE34A893A93D61AFF204018A2801010114C81BDD2EF2B3541B6DFB667EDA3843297B42C3E3 T:1}"
D[04-04|09:59:29.819] Received bytes                               module=p2p validator=5 peer=pipe chID=32 msgBytes="\u0016\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd"
D[04-04|09:59:29.819] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=32 bytes="\u0016\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd"
D[04-04|09:59:29.819] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:01000000000000001300010000 T:1}"
D[04-04|09:59:29.819] Received bytes                               module=p2p validator=5 peer=pipe chID=32 msgBytes="\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0013\u0000\u0001\u0000\u0000"
D[04-04|09:59:29.820] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=32 bytes="\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0013\u0000\u0001\u0000\u0000"
D[04-04|09:59:29.820] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:15000000000000001200020104 T:1}"
D[04-04|09:59:29.820] Received bytes                               module=p2p validator=5 peer=pipe chID=32 msgBytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0002\u0001\u0004"
D[04-04|09:59:29.820] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=32 bytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0002\u0001\u0004"
D[04-04|09:59:29.820] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{20:15000000000000001200020101 T:1}"
D[04-04|09:59:29.820] Received bytes                               module=p2p validator=5 peer=pipe chID=32 msgBytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0002\u0001\u0001"
D[04-04|09:59:29.820] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=32 bytes="\u0015\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0002\u0001\u0001"
D[04-04|09:59:29.820] Read Msg Packet                              module=p2p validator=5 peer=pipe conn=MConn{pipe} packet="MsgPacket{22:14010114F2BAB045DD6C33B349872067C4D90A17EC4DB85B01040000000000000011001522333DC4894100020114629B77BE310C25AE34A893A93D61AFF204018A2801010114C81BDD2EF2B3541B6DFB667EDA3843297B42C3E301E647FAD4C25366DB07A438217D720D48A8C28AA75F6558865A363EFAD705DB4D7C28FA15C2947811CDB1259F3AD87FE146AFC5B5F3CA3193CCCD7BDA92198906 T:1}"
D[04-04|09:59:29.820] Received bytes                               module=p2p validator=5 peer=pipe chID=34 msgBytes="\u0014\u0001\u0001\u0014\ufffd\ufffd\ufffdE\ufffdl3\ufffdI\ufffd g\ufffd\ufffd\n\u0017\ufffdM\ufffd[\u0001\u0004\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0015\"3=?~IA\u0000\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd\u0001\ufffdG\ufffd\ufffd\ufffdSf\ufffd\u0007\ufffd8!}r\rH\ufffd?~J\ufffd_eX\ufffdZ6>\ufffd\ufffd\u0005\ufffdM|(\ufffd\u0015?~Tx\u0011ͱ%\ufffd:\ufffd^?\ufffdF\ufffdŵ\ufffd\ufffd1\ufffd\ufffd\ufffd{?~R\u0019\ufffd\u000
06"
D[04-04|09:59:29.821] Receive                                      module=consensus validator=5 src="Peer{MConn{pipe} bc2437e463a3fe274ab12587b1372254feafde41 in}" chId=34 bytes="\u0014\u0001\u0001\u0014\ufffd\ufffd\ufffdE\ufffdl3\ufffdI\ufffd g\ufffd\ufffd\n\u0017\ufffdM\ufffd[\u0001\u0004\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u0015\"3=?~IA\u0000\u0002\u0001\u0014b\ufffdw\ufffd1\u000c%\ufffd4\ufffd\ufffd\ufffd=a\ufffd\ufffd\u0004\u0001\ufffd(\u0001\u0001\u0001\u0014\ufffd\u001b\ufffd.\ufffd\ufffdT\u001bm\ufffdf~\ufffd8C){B\ufffd\ufffd\u0001\ufffdG\ufffd\ufffd\ufffdSf\ufffd\u0007\ufffd8!}r\rH\ufffd?~J\ufffd_eX\ufffdZ6>\ufffd\ufffd\u0005\ufffdM|(\ufffd\u0015?~Tx\u0011ͱ%\ufffd:\ufffd^?\uffff
dF\ufffdŵ\ufffd\ufffd1\ufffd\ufffd\ufffd{?~R\u0019\ufffd\u0006"
I[04-04|09:59:29.821] Stopping baseWAL                             module=consensus validator=5 wal=/home/ubuntu/.tendermint_test/consensus_val_set_changes_test_5/data/cs.wal/wal impl=baseWAL
D[04-04|09:59:29.821] Send Pong                                    module=p2p validator=5 peer=pipe
D[04-04|09:59:29.821] Flush                                        module=p2p validator=5 peer=pipe conn=MConn{pipe}
E[04-04|09:59:29.821] MConnection flush failed                     module=p2p validator=5 peer=pipe err="io: read/write on closed pipe"

--- PASS: TestWALSearchForEndHeight (0.40s)
FAIL
I[04-04|09:59:48.838] Executed block                               height=6 validTxs=0 invalidTxs=0
I[04-04|09:59:48.838] Committed state                              height=6 txs=0 appHash=0000000000000000
D[04-04|09:59:48.839] WAL already stopped. Not writing message     wal_generator=wal_generator msg="{Height:7 Round:0 Step:RoundStepNewHeight RoundState:RoundState{\n  H:7 R:0 S:RoundStepNewHeight\n  StartTime:     2018-04-04 09:59:48.831517861 +0000 UTC m=+110.428351348\n  CommitTime:    2018-04-04 09:59:48.821517861 +0000 UTC m=+110.418351348\n  Validators:    ValidatorSet{\n      Proposer: Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n      Validators:\n        Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n    }\n  Proposal:      <nil>\n  ProposalBlock: nil-PartSet nil-Block\n  LockedRound:   0\n  LockedBlock:   nil-PartSet nil-Block\n  ValidRound:   0\n  ValidBlock:   nil-PartSet nil-Block\n  Votes:         HeightVoteSet{H:7 R:0~0\n      VoteSet{H:7 R:0 T:1 +2/3:<nil> BA{1:_} map[]}\n      VoteSet{H:7 R:0 T:2 +2/3:<nil> BA{1:_} map[]}\n    }\n  LastCommit:    VoteSet{H:6 R:0 T:2 +2/3:2026AFBA8501C4CB88BC7EBC4B4BC8F6955DD8FC:1:27279E1ADADB BA{1:X} map[]}\n  LastValidators:ValidatorSet{\n      Proposer: Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n      Validators:\n        Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n    }\n}}"
I[04-04|09:59:48.839] enterNewRound(7/0). Current: 7/0/RoundStepNewHeight wal_generator=wal_generator
I[04-04|09:59:48.839] enterPropose(7/0). Current: 7/0/RoundStepNewRound wal_generator=wal_generator
D[04-04|09:59:48.840] This node is a validator                     wal_generator=wal_generator
I[04-04|09:59:48.840] enterPropose: Our turn to propose            wal_generator=wal_generator proposer=D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 privValidator="PrivValidator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 LH:6, LR:0, LS:3}"
I[04-04|09:59:48.845] Signed proposal                              wal_generator=wal_generator height=7 round=0 proposal="Proposal{7/0 1:814C1A1F8DE5 (-1,:0:000000000000) {/E01D935AE0C6.../} @ 2018-04-04T09:59:48.841Z}"
D[04-04|09:59:48.847] Signed proposal block: Block{
  Header{
    ChainID:        tendermint_test
    Height:         7
    Time:           2018-04-04 09:59:48.840155991 +0000 UTC m=+110.436989357
    NumTxs:         0
    TotalTxs:       0
    LastBlockID:    2026AFBA8501C4CB88BC7EBC4B4BC8F6955DD8FC:1:27279E1ADADB
    LastCommit:     B7F6061D931C88F4986D81FC045846E70177EF5D
    Data:
    Validators:     7297262C6CD96190E46846C9A0DE1227E76077CF
    App:            0000000000000000
    Conensus:       0B8CEF95EC57AC2D96038FD0AE3901C14FAE8E73
    Results:
    Evidence:
  }#1564CACC287A48695611031320D82292A18578D3
  Data{

  }#
  Data{

  }#
  Commit{
    BlockID:    2026AFBA8501C4CB88BC7EBC4B4BC8F6955DD8FC:1:27279E1ADADB
    Precommits: Vote{0:D028C9981F7A 6/00/2(Precommit) 2026AFBA8501 {/EC49A4BFDCED.../} @ 2018-04-04T09:59:48.809Z}
  }#B7F6061D931C88F4986D81FC045846E70177EF5D
}#1564CACC287A48695611031320D82292A18578D3 wal_generator=wal_generator
D[04-04|09:59:48.847] WAL already stopped. Not writing message     wal_generator=wal_generator msg="{Height:7 Round:0 Step:RoundStepPropose RoundState:RoundState{\n  H:7 R:0 S:RoundStepPropose\n  StartTime:     2018-04-04 09:59:48.831517861 +0000 UTC m=+110.428351348\n  CommitTime:    2018-04-04 09:59:48.821517861 +0000 UTC m=+110.418351348\n  Validators:    ValidatorSet{\n      Proposer: Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n      Validators:\n        Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n    }\n  Proposal:      <nil>\n  ProposalBlock: nil-PartSet nil-Block\n  LockedRound:   0\n  LockedBlock:   nil-PartSet nil-Block\n  ValidRound:   0\n  ValidBlock:   nil-PartSet nil-Block\n  Votes:         HeightVoteSet{H:7 R:0~1\n      VoteSet{H:7 R:0 T:1 +2/3:<nil> BA{1:_} map[]}\n      VoteSet{H:7 R:0 T:2 +2/3:<nil> BA{1:_} map[]}\n      VoteSet{H:7 R:1 T:1 +2/3:<nil> BA{1:_} map[]}\n      VoteSet{H:7 R:1 T:2 +2/3:<nil> BA{1:_} map[]}\n    }\n  LastCommit:    VoteSet{H:6 R:0 T:2 +2/3:2026AFBA8501C4CB88BC7EBC4B4BC8F6955DD8FC:1:27279E1ADADB BA{1:X} map[]}\n  LastValidators:ValidatorSet{\n      Proposer: Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n      Validators:\n        Validator{D028C9981F7A87F3093672BF0D5B0E2A1B3ED456 {PubKeyEd25519{3B3069C422E19688B45CBFAE7BB009FC0FA1B1EA86593519318B7214853803C8}} VP:10 A:0}\n    }\n}}"
D[04-04|09:59:48.848] WAL already stopped. Not writing message     wal_generator=wal_generator msg="{Msg:[Proposal Proposal{7/0 1:814C1A1F8DE5 (-1,:0:000000000000) {/E01D935AE0C6.../} @ 2018-04-04T09:59:48.841Z}] PeerID:}"
D[04-04|09:59:48.851] WAL already stopped. Not writing message     wal_generator=wal_generator msg="{Msg:[BlockPart H:7 R:0 P:Part{#0\n  Bytes: 0101010F7465...\n  Proof: SimpleProof{\n    Aunts: []\n  }\n}] PeerID:}"
I[04-04|09:59:48.853] Received complete proposal block             wal_generator=wal_generator height=7 hash=1564CACC287A48695611031320D82292A18578D3
I[04-04|09:59:48.853] enterPrevote(7/0). Current: 7/0/RoundStepPropose wal_generator=wal_generator
coverage: 75.6% of statements
exit status 1
FAIL    github.com/tendermint/tendermint/consensus      110.530s
Makefile:82: recipe for target 'test_cover' failed

--- a/consensus/reactor.go
+++ b/consensus/reactor.go
@@ -1087,6 +1087,9 @@ func (ps *PeerState) ensureVoteBitArrays(height int64, numValidators int) {
 // It returns the total number of votes (1 per block). This essentially means
 // the number of blocks for which peer has been sending us votes.
 func (ps *PeerState) RecordVote(vote *types.Vote) int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
 	if ps.stats.lastVoteHeight >= vote.Height {
 		return ps.stats.votes
 	}
@@ -1098,13 +1101,20 @@ func (ps *PeerState) RecordVote(vote *types.Vote) int {
 // VotesSent returns the number of blocks for which peer has been sending us
 // votes.
 func (ps *PeerState) VotesSent() int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
 	return ps.stats.votes
 }
 
-// RecordVote updates internal statistics for this peer by recording the block part.
-// It returns the total number of block parts (1 per block). This essentially means
-// the number of blocks for which peer has been sending us block parts.
+// RecordBlockPart updates internal statistics for this peer by recording the
+// block part. It returns the total number of block parts (1 per block). This
+// essentially means the number of blocks for which peer has been sending us
+// block parts.
 func (ps *PeerState) RecordBlockPart(bp *BlockPartMessage) int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
 	if ps.stats.lastBlockPartHeight >= bp.Height {
 		return ps.stats.blockParts
 	}
@@ -1117,6 +1127,9 @@ func (ps *PeerState) RecordBlockPart(bp *BlockPartMessage) int {
 // BlockPartsSent returns the number of blocks for which peer has been sending
 // us block parts.
 func (ps *PeerState) BlockPartsSent() int {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
 	return ps.stats.blockParts
 }
 

